firedrake/ensemble/ensemble_functionspace.py

```diff
@@ -0,0 +1,242 @@
+from firedrake.petsc import PETSc
+from firedrake.functionspace import MixedFunctionSpace
+from ufl.duals import is_primal, is_dual
+from pyop2.mpi import internal_comm
+from functools import cached_property
+from itertools import chain
+
+
+def _is_primal_or_dual(local_spaces, ensemble):
+    """
+    We need to test primal or dual collectively over all ensemble
+    ranks otherwise some may fail when others pass.
```

Comment:
Should be in the imperative tone. E.g.: "Return whether all spaces in an ensemble are primal or dual."
The motivation can be written below (with a gap of one line)